pkg/devfile/parser/data/v2/components.go

```diff
@@ -10,6 +10,52 @@ func (d *DevfileV2) GetComponents() []v1.Component {
 	return d.Components
 }
 
+// GetDevfileContainerComponents iterates through the components in the devfile and returns a list of devfile container components
+func (d *DevfileV2) GetDevfileContainerComponents() []v1.Component {
+	var components []v1.Component
+	for _, comp := range d.GetComponents() {
+		if comp.Container != nil {
+			components = append(components, comp)
+		}
+	}
+	return components
+}
+
+// GetDevfileVolumeComponents iterates through the components in the devfile and returns a list of devfile volume components
+func (d *DevfileV2) GetDevfileVolumeComponents() []v1.Component {
+	var components []v1.Component
+	for _, comp := range d.GetComponents() {
+		if comp.Volume != nil {
+			components = append(components, comp)
+		}
+	}
+	return components
+}
+
+// GetPortExposure iterate through all endpoints and returns the highest exposure level of all TargetPort.
+// exposure level: public > internal > none
+func (d *DevfileV2) GetPortExposure() map[int]v1.EndpointExposure {
```

Comment:
not sure about this one in this package? is it intended to be directly consumed by library users? i saw  generators code using this

Reply:
True, it is used by the generator code atm. But the reasoning behind why it was moved to the parser pkg is because the func returns a map with type devfile type `v1.EndpointExposure`. We wanted to keep the generator pkg for only Kubernetes type functions.

Any thoughts?

Reply:
exposed function should return kube objects, but if it is not intended to be used by consumer, we can move this into generator/utils

Reply:
Updated